src/API.jl

```diff
@@ -895,30 +895,72 @@ end
 precompile() = precompile(Context())
 function precompile(ctx::Context)
     printpkgstyle(ctx, :Precompiling, "project...")
-
-    pkgids = [Base.PkgId(uuid, name) for (name, uuid) in ctx.env.project.deps if !is_stdlib(uuid)]
+    
+    num_tasks = parse(Int, get(ENV, "JULIA_NUM_PRECOMPILE_TASKS", string(Sys.CPU_THREADS + 1)))
+    parallel_limiter = Base.Semaphore(num_tasks)
+    
+    man = Pkg.Types.read_manifest(ctx.env.manifest_file)
+    pkgids = [Base.PkgId(first(dep), last(dep).name) for dep in man if !Pkg.Operations.is_stdlib(first(dep))]
```

Comment:
I know this filtering was here before, but why is it necessary to filter out stdlibs?
```
(tmp.9L2rmMdEXO) pkg> st
Status `/tmp/tmp.9L2rmMdEXO/Project.toml`
  [37e2e46d] LinearAlgebra

(tmp.9L2rmMdEXO) pkg> precompile
Precompiling project...

julia> using LinearAlgebra
[ Info: Precompiling LinearAlgebra [37e2e46d-f89d-539d-b4ee-838fcccc9c8e]
```

Reply:
Aren't stdlib's always going to be precompiled already, and if you're dev-ing them they'd need to have their uuid removed, so wouldn't identify as stdlibs in that check?

Reply:
Right, forgot to say that this is when you compile Julia without them in the sysimg. Perhaps we can instead filter based on if the package is already loaded? That should work for both regular packages and stdlibs. If it is a stdlib that is in the sysimg it doesn't need to precompile, and if it is a regular package that is already loaded in the session it is probably just precompiled from the `using`?

Reply:
But what if they're loaded, and in need of recompiling? Perhaps the filter just isn't needed at all?

Reply:
Yea I am not sure what happens if you try to precompile stdlibs that are loaded though? Since no precompiles files exist, will that spend time on precompiling them anyway? At least we can add the filter I suggested to the stdlibs.

Reply:
You can update a package after importing it. In this case, I think `Pkg.precompile` should re-compile it and all its downstream packages. So, I don't think it's a good strategy to check if it's loaded or not. Maybe it's better to have an API `Base.in_sysimage(pkg)`?

Reply:
That sounds good. Do you have a design of how that might work?

Reply:
Maybe the easiest way to do it is to copy `keys(loaded_modules)` in `Base.__init__` to some constant?

Reply:
> You can update a package after importing it.

Yea but not really with stdlibs. If we for stdlibs check if they are loaded or not, then the only thing we potentially miss is if someone has (i) loaded the stdlib and (ii) changed version of it which is a really niche thing to do, so I think it is fine.

Reply:
Before I saw your last message @fredrikekre I got @tkf's last suggestion working. Should I PR that?

Reply:
IMO that seems overly complicated for a corner case that will not ever happen really.

Reply:
Ok, got the alternative working with this. I'll convert the relevant part of the code to work off `PkgId[]` rather than `UUID[]` to make this cleaner. But it works as a prototype
```
function is_stdlib_and_loaded(uuid::UUID)
    return Pkg.Operations.is_stdlib(uuid) && (uuid in map(x->key(x).uuid, module_keys))
end
```

Reply:
https://github.com/JuliaLang/Pkg.jl/pull/2021 updated with this now (the PkgId version)

Reply:
I thought the stdlib check was for some kind of optimization (launching `julia` takes a few hundreds of ms even if you don't do anything). So, I think the correct predicate here is "is it in sysimage?" than "is it a stdlib?" since non-stdlib packages can be in sysimage and there is no point in calling `compilecache` for them.  This also covers the exotic situation where some stdlibs are not in sysimage.

I think `is_stdlib_and_loaded` is better than nothing. But I feel it's a bit half-way solution if my assumption (the stdlib check was optimization) is correct.

Reply:
Thats true, I didn't think about regular packages in the sysimg. But perhaps https://github.com/JuliaLang/Pkg.jl/pull/2018#discussion_r490555795 is a good enough approximation of that? It seems pretty strange to (i) load a dependency, (ii) update its version, (iii) pkg> precompile, (iv) restart Julia and expect everything to be precompiled?

Reply:
Although #2021 is looking good, I do like the properness of `in_sysimage`. It explains exactly why we'd always want to skip. I'll prepare a PR

Reply:
> It seems pretty strange to (i) load a dependency, (ii) update its version, (iii) pkg> precompile, (iv) restart Julia and expect everything to be precompiled?

@fredrikekre Hmm... That was my expectation, actually. I generally expect `pkg> $cmd` and `shell> jlpkg $cmd` to be identical (when a project is not activated). Anyway, what do you think about #2021 + https://github.com/JuliaLang/julia/pull/37652? I think `in_sysimage` is simple enough and nice to have.